### PR TITLE
telescope.harpoon: init 

### PR DESCRIPTION
**Pull request by hmajid2301**
Added telescope extension: [harpoon](https://github.com/ThePrimeagen/harpoon#telescope-support)

### CODE REVIEW

**Review by GaetanLepage — COMMENTED** [review]
Shouldn't we check/require that harpoon (the plugin) is enabled in this case ?

**Comment by hmajid2301** [comment]
Yes, whats the best way to do that ? `config.plugins.utils.harpoon.enable` ?

**Comment by traxys** [comment]
I feel we should add this in the harpoon plugin directly, with something like `enableTelescope = mkEnableOption "telescope intergration for harpoon"`.

Though we should raise a warning if telescope is not enabled 
(example for treesitter:
```nix
      warnings = mkIf (!config.plugins.treesitter.enable) [
        "Nixvim: ts-context-commentstring needs treesitter to function as intended"
      ];
```)

**Comment by hmajid2301** [comment]
So I guess we want something like:

```nix
  mkIf cfg.enableTelescope {
  warnings = mkIf (!config.plugins.telescope.enable) [
    "Nixvim: The harpoon telescope integration needs telescope to function as intended"
  ];

  extraConfigLua = ''
   require("telescope").load_extension('harpoon')
  '';
};
```

Where do I put this ? I dont think I can put in the harpoon block i.e. nested `mkIf`, shall I do something like this with mkMerge ? 

```
config = mkIf cfg.enable (mkMerge [
   (mkIf condition1 {
      ...
   })
   (mkIf condition2 {
      ...
   })
   ....
]);
```

**Comment by traxys** [comment]
Hm not sure, you can try to do nested `mkIf`. For warnings you can do a `lib.optional` I think?

**Comment by hmajid2301** [comment]
Trying to do:

```
 mkIf cfg.enable {
      mkIf cfg.enableTelescope {
      warnings = mkIf (!config.plugins.telescope.enable) [
        "Nixvim: The harpoon telescope integration needs telescope to function as intended"
      ];

      extraConfigLua = ''
        require("telescope").load_extension('harpoon')
      '';
    };
```

gives me a syntax error. `syntax error, unexpected ID, expecting '.' or '=' `

**Comment by traxys** [comment]
Yeah you want be able to merge them like this.

You can do the following: 
```nix
mkIf cfg.enable  {
      warnings = mkIf (!config.plugins.telescope.enable && cfg.enableTelescope) [
        "Nixvim: The harpoon telescope integration needs telescope to function as intended"
      ];
      
      extraConfigLua = let
        telescopeCfg = ''require(....).load....'';
      in ''
          //curent harpoon content of extraConfigLua
          ${if cfg.enableTelescope then telescopeCfg else ""}
      '';
}
```

In the harpoon file.

Maybe more of an assertion than a warning, as the config will error.
```nix
assertions = [
{
   assertion = /* boolean */;;
  message = '' .... '';
}
];
```

**Comment by hmajid2301** [comment]
I believe I have made the changes!

File: plugins/utils/harpoon.nix [hunk 1]
```diff
@@ -198,10 +200,21 @@ in {
       // cfg.extraOptions;
   in
     mkIf cfg.enable {
+      warnings = mkIf (!config.plugins.telescope.enable && cfg.enableTelescope) [
+        "Nixvim: The harpoon telescope integration needs telescope to function as intended"
+      ];
```
Comment (traxys) [inline plugins/utils/harpoon.nix:205]: Please use an assertion instead of a warning, see my previous comment about those

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/harpoon.nix (hunk 1 above)
Comment (hmajid2301) [reply plugins/utils/harpoon.nix:205]: apologies i forgot to make that change!

File: tests/test-sources/plugins/utils/harpoon.nix [hunk 2]
```diff
@@ -15,6 +15,7 @@
     plugins.harpoon = {
       enable = true;
 
+      enableTelescope = true;
```
Comment (traxys) [inline tests/test-sources/plugins/utils/harpoon.nix:18]: You are going to need to enable telescope in that test, else the assertion will trigger :P

**Review by traxys — CHANGES_REQUESTED** [review]

**Comment by traxys** [comment]
I think you need that operator:
https://nixos.org/manual/nix/stable/language/operators#logical-implication

File: tests/test-sources/plugins/utils/harpoon.nix (hunk 2 above)
Comment (hmajid2301) [reply tests/test-sources/plugins/utils/harpoon.nix]: yeh i disabled it but maybe it makes more sense to test with it enabled

**Comment by traxys** [comment]
> > I think you need that operator:
> > https://nixos.org/manual/nix/stable/language/operators#logical-implication
> 
> Do you mean in the assertion bit ? `!config.plugins.telescope.enable -> cfg.enableTelescope`

Yes, but I think you need to revert it. cfg.enableTelescope implies that config.plugins.telescope.enable

**Comment by hmajid2301** [comment]
> > > I think you need that operator:
> > > https://nixos.org/manual/nix/stable/language/operators#logical-implication
> > 
> > 
> > Do you mean in the assertion bit ? `!config.plugins.telescope.enable -> cfg.enableTelescope`
> 
> Yes, but I think you need to revert it. cfg.enableTelescope implies that config.plugins.telescope.enable

Ahh ofc yes, basic logic :facepalm: !

**Comment by hmajid2301** [comment]
> Can you also write a test where enableTelescope is false ? That code could be error prone in the future.

Yes ofc!

**Review by traxys — APPROVED** [review]

**Pull request merged**